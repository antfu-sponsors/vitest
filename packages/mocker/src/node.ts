export { automockModule, automockPlugin } from './plugin'
export { findMockRedirect } from './redirect'
