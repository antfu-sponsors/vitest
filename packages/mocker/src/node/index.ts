export { automockModule, automockPlugin } from './automockPlugin'
export { findMockRedirect } from './redirect'
export { hoistMocksPlugin, hoistMocks } from './hoistMocksPlugin'
export { ServerMockResolver } from './resolver'
export { dynamicImportPlugin } from './dynamicImportPlugin'
export { interceptorPlugin } from './interceptorPlugin'
export { mockerPlugin } from './mockerPlugin'

export type {
  ServerMockResolution,
  ServerIdResolution,
  ServerResolverOptions,
} from './resolver'
export type { AutomockPluginOptions } from './automockPlugin'
export type { HoistMocksPluginOptions, HoistMocksResult } from './hoistMocksPlugin'
export type { InterceptorPluginOptions } from './interceptorPlugin'
