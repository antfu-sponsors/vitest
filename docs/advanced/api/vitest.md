---
outline: deep
---

# Vitest

Vitest instance requires the current test mode. It can be either:

- `test` when running runtime tests
- `benchmark` when running benchmarks

## mode

### test

Test mode will only call functions inside `test` or `it`, and throws an error when `bench` is encountered. This mode uses `include` and `exclude` options in the config to find test files.

### benchmark

Benchmark mode calls `bench` functions and throws an error, when it encounters `test` or `it`. This mode uses `benchmark.include` and `benchmark.exclude` options in the config to find benchmark files.

## start

```ts
function start(filters: string[]): Promise<TestRunResult>
```

You can start running tests or benchmarks with `start` method. You can pass an array of strings to filter test files.

## config

The root (or global) config. If workspace feature is enabled, projects will reference this as `globalConfig`.

::: warning
This is Vitest config, it doesn't extend _Vite_ config. It only has resolved values from the `test` property.
:::

## vite

This is a global [`ViteDevServer`](https://vite.dev/guide/api-javascript#vitedevserver).

## state

::: warning
Public `state` is an experimental API. Breaking changes might not follow SemVer, please pin Vitest's version when using it.
:::

Global state stores information about the current tests. It uses the same API from `@vitest/runner` by default, but we recommend using the [Reported API](/advanced/reporters#reported-tasks) instead by calling `state.getReportedEntity()` on the `@vitest/runner` API:

```ts
const task = vitest.state.idMap.get(taskId) // old API
const testCase = vitest.state.getReportedEntity(task) // new API
```

In the future, the old API won't be exposed anymore.

## snapshot

The global snapshot manager. Vitest keeps track of all snapshots using the `snapshot.add` method.

You can get the latest summary of snapshots via the `vitest.snapshot.summay` property.

## cache

Cache manager that stores information about latest test results and test file stats. In Vitest itself this is only used by the default sequencer to sort tests.

## ready

Vitest needs to be resolved with the Vite server to be properly initialized. If the `Vitest` instance was created manually, you might need to check the `ready` status before accessing the `vite`, `state`, `cache`, `config`, and `snapshot` properties; otherwise, they will throw an error in the getter.

In normal circumstances, you would never call this method because `createVitest` and `startVitest` return already resolved Vitest instance.

## getRootTestProject

This returns the root test project. The root project generally doesn't run any tests and is not included in `vitest.projects` unless the user explicitly includes the root config in their workspace.

The primary goal of the root project is to setup the global config. In fact, `rootProject.config` references `rootProject.globalConfig` and `vitest.config` directly.

## provide

Vitest exposes `provide` method which is a shorthand for `vitest.getRootTestProject().provide`. With this method you can pass down values from the main thread to tests. All values are checked with `structuredClone` before they are stored, but the values themselves are not cloned.

To recieve the values in the test, you need to import `inject` method from `vitest` entrypont:

```ts
import { inject } from 'vitest'
const port = inject('wsPort') // 3000
```

For better type safety, we encourage you to augment the type of `ProvidedContext`:

```ts
import { createVitest } from 'vitest/node'

const vitest = await createVitest('test', {
  watch: false,
})
vitest.provide('wsPort', 3000)

declare module 'vitest' {
  export interface ProvidedContext {
    wsPort: number
  }
}
```

::: warning
Technically, `provide` is a method of [`TestProject`](#testproject), so it is limited to the specific project. However, all projects inherit the values from the core project which makes `vitest.provide` universal way of passing down values to tests.
:::

## getProvidedContext

This returns the root context object. This is a shorthand for `vitest.getRootTestProject().getProvidedContext`.

## getProjectByName

This method returns the project by its name. Simillar to calling `vitest.projects.find`.

::: warning
In case the project doesn't exist, this method will return the root project - make sure to check the names again if you need to make sure the project you are looking for is the one returned.
:::

## globTestSpecifications

This method constructs new [test specifications](#testspecification) by collecting every test in all projects with [`project.globTestFiles`](#globtestfiles). It accepts string filters to match the test files.

::: warning
As of Vitest 2.2.0, it's possible to have multiple test specifications with the same module ID (file path) if `poolMatchGlob` has several pools or if `typecheck` is enabled. This possibility will be removed in Vitest 3.
:::

```ts
const specifications = await vitest.globTestSpecifications(['my-filter'])
// [TestSpecification{ moduleId: '/tests/my-filter.test.ts' }]
console.log(specifications)
```

## mergeReports
## collect
## listFiles
## start
## init

## getModuleSpecifications

Returns a list of test specifications related to the module ID. The ID should already be resolved to an absolute file path. If ID doesn't match `include` or `includeSource` patterns, the returned array will be empty.

::: warning
As of Vitest 2.2.0, this method uses a cache to check if the file is a test. To make sure that the cache is not empty, call `globTestSpecifications` at least once.
:::

## runTestSpecifications
## rerunTestSpecifications
## collectTests
## cancelCurrentRun
## updateSnapshot
## invalidateFile
## close
## exit

## shouldKeepServer
## onServerRestart
## onCancel
## onClose
## onTestsRerun
## onFilterWatchedSpecification
