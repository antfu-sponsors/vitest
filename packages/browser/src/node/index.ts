import type { WorkspaceProject } from 'vitest/node'
import type { Plugin } from 'vitest/config'
import { createViteLogger, createViteServer } from 'vitest/node'
import c from 'tinyrainbow'
import { version } from '../../package.json'
import { setupBrowserRpc } from './rpc'
import { BrowserServer } from './server'
import BrowserPlugin from './plugin'

export type { BrowserServer } from './server'
export { createBrowserPool } from './pool'

export async function createBrowserServer(
  project: WorkspaceProject,
  configFile: string | undefined,
  prePlugins: Plugin[] = [],
  postPlugins: Plugin[] = [],
) {
  if (project.ctx.version !== version) {
    project.ctx.logger.warn(
      c.yellow(
        `Loaded ${c.inverse(c.yellow(` vitest@${project.ctx.version} `))} and ${c.inverse(c.yellow(` @vitest/browser@${version} `))}.`
        + '\nRunning mixed versions is not supported and may lead into bugs'
        + '\nUpdate your dependencies and make sure the versions match.',
      ),
    )
  }

  const server = new BrowserServer(project, '/')

  const configPath = typeof configFile === 'string' ? configFile : false

  const logLevel = (process.env.VITEST_BROWSER_DEBUG as 'info') ?? 'info'

  const logger = createViteLogger(logLevel)

  const vite = await createViteServer({
    ...project.options, // spread project config inlined in root workspace config
    base: '/',
    logLevel,
    customLogger: {
      ...logger,
      info(msg, options) {
        logger.info(msg, options)
        if (msg.includes('optimized dependencies changed. reloading')) {
          logger.warn(
            [
              c.yellow(`\n${c.bold('[vitest]')} Vite unexpectedly reloaded a test. This may cause tests to fail, lead to flaky behaviour or duplicated test runs.\n`),
              c.yellow(`For a stable experience, please add mentioned dependencies to your config\'s ${c.bold('\`optimizeDeps.include\`')} field manually.\n\n`),
            ].join(''),
          )
        }
      },
    },
    mode: project.config.mode,
    configFile: configPath,
    // watch is handled by Vitest
    server: {
      hmr: false,
      watch: null,
    },
    plugins: [
      ...prePlugins,
      ...(project.options?.plugins || []),
      BrowserPlugin(server),
      ...postPlugins,
    ],
  })

  await vite.listen()

  setupBrowserRpc(server)

  return server
}
