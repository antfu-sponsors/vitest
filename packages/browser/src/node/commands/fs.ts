import fs, { promises as fsp } from 'node:fs'
import { dirname, resolve } from 'node:path'
import { isFileServingAllowed } from 'vite'
import type { WorkspaceProject } from 'vitest/node'
import type { BrowserCommand } from '../types'

export interface FsOptions {
  encoding?: BufferEncoding
  flag?: string | number
}

function assertFileAccess(path: string, project: WorkspaceProject) {
  const resolvedPath = resolve(path)
  if (!isFileServingAllowed(resolvedPath, project.server) && !isFileServingAllowed(resolvedPath, project.ctx.server))
    throw new Error(`Access denied to "${resolvedPath}". See Vite config documentation for "server.fs": https://vitejs.dev/config/server-options.html#server-fs-strict.`)
}

export const readFile: BrowserCommand<[string, BufferEncoding | FsOptions]> = async ([path, options], { project }) => {
  assertFileAccess(path, project)
  return fsp.readFile(path, options).catch(() => null)
}

export const writeFile: BrowserCommand<[string, string, BufferEncoding | FsOptions & { mode?: number | string }]> = async ([path, data, options], { project }) => {
  assertFileAccess(path, project)
  const dir = dirname(path)
  if (!fs.existsSync(dir))
    await fsp.mkdir(dir, { recursive: true })
  await fsp.writeFile(path, data, options).catch(() => null)
}

export const removeFile: BrowserCommand<[string]> = async ([path], { project }) => {
  assertFileAccess(path, project)
  await fsp.rm(path).catch(() => null)
}
