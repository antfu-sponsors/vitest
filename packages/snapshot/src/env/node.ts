import { existsSync, promises as fs } from 'node:fs'
import { basename, dirname, isAbsolute, join, resolve } from 'pathe'
import type { SnapshotEnvironment } from '../types'

export class NodeSnapshotEnvironment implements SnapshotEnvironment {
  getVersion(): string {
    return '1'
  }

  getHeader(): string {
    return `// Snapshot v${this.getVersion()}`
  }

  async resolveRawPath(testPath: string, rawPath: string) {
    return isAbsolute(rawPath)
      ? rawPath
      : resolve(dirname(testPath), rawPath)
  }

  async resolvePath(filepath: string): Promise<string> {
    return join(
      join(
        dirname(filepath), '__snapshots__'),
      `${basename(filepath)}.snap`,
    )
  }

  async prepareDirectory(filepath: string): Promise<void> {
    await fs.mkdir(filepath, { recursive: true })
  }

  async saveSnapshotFile(filepath: string, snapshot: string): Promise<void> {
    await fs.writeFile(filepath, snapshot, 'utf-8')
  }

  async readSnapshotFile(filepath: string): Promise<string | null> {
    if (!existsSync(filepath))
      return null
    return fs.readFile(filepath, 'utf-8')
  }

  async removeSnapshotFile(filepath: string): Promise<void> {
    if (existsSync(filepath))
      await fs.unlink(filepath)
  }
}
