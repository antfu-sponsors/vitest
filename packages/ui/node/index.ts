import { fileURLToPath } from 'node:url'
import { readFile } from 'node:fs/promises'
import { resolve } from 'pathe'
import sirv from 'sirv'
import type { Plugin } from 'vite'
import type { Vitest } from 'vitest'
import { resolveCoverageFolder, transformCoverageEntryPoint } from './utils'

export default (ctx: Vitest): Plugin => {
  return <Plugin>{
    name: 'vitest:ui',
    apply: 'serve',
    configureServer(server) {
      const uiOptions = ctx.config
      const base = uiOptions.uiBase
      const coverageFolder = resolveCoverageFolder(ctx)
      const coveragePath = coverageFolder ? coverageFolder[1] : undefined
      if (coveragePath && base === coveragePath) {
        throw new Error(
          `The ui base path and the coverage path cannot be the same: ${base}, change coverage.reportsDirectory`,
        )
      }

      if (coverageFolder) {
        const root = resolve(fileURLToPath(import.meta.url), '../')
        const base = readFile(resolve(root, '../istanbul-base.css'))
        const prettify = readFile(resolve(root, '../istanbul-prettify.css'))
        server.middlewares.use(coveragePath!, async (req, res, next) => {
          if (!req.url) {
            return next()
          }

          const pathname = new URL(req.url, 'http://localhost').pathname
          if (pathname !== '/' && pathname !== '/index.html' && pathname !== '/base.css' && pathname !== '/prettify.css' && !pathname.endsWith('.html')) {
            return next()
          }

          if (pathname === '/base.css') {
            res.setHeader('Content-Type', 'text/css')
            res.setHeader('Cache-Control', 'public,max-age=0,must-revalidate')
            res.end(await base)
            // for testing purposes: don't remove it
            // res.end(await readFile(resolve(fileURLToPath(import.meta.url), '../../istambul-base.css')))
            return
          }

          if (pathname === '/prettify.css') {
            res.setHeader('Content-Type', 'text/css')
            res.setHeader('Cache-Control', 'public,max-age=0,must-revalidate')
            res.end(await prettify)
            // for testing purposes: don't remove it
            // res.end(await readFile(resolve(fileURLToPath(import.meta.url), '../../istambul-prettify.css')))
            return
          }

          const file = pathname.endsWith('.html')
            ? pathname[0] === '/' ? pathname.slice(1) : pathname
            : 'index.html'

          const indexHtml = await readFile(resolve(coverageFolder[0], file), 'utf-8')
          res.setHeader('Content-Type', 'text/html')
          res.setHeader('Cache-Control', 'public,max-age=0,must-revalidate')
          res.end(transformCoverageEntryPoint(indexHtml))
        })
        server.middlewares.use(
          coveragePath!,
          sirv(coverageFolder[0], {
            single: true,
            dev: true,
            setHeaders: (res) => {
              res.setHeader(
                'Cache-Control',
                'public,max-age=0,must-revalidate',
              )
            },
          }),
        )
      }

      const clientDist = resolve(fileURLToPath(import.meta.url), '../client')
      server.middlewares.use(
        base,
        sirv(clientDist, {
          single: true,
          dev: true,
        }),
      )
    },
  }
}
