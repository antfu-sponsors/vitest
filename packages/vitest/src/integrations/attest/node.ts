import type { TestProject } from '../../node/project'
import { execFileSync } from 'node:child_process'
import { mkdirSync } from 'node:fs'

async function setup() {
  // TODO: programatic API
  mkdirSync('.attest/assertions', { recursive: true })
  execFileSync('attest', ['precache', '.attest/assertions/typescript.json'], {
    stdio: 'inherit',
    env: {
      ...process.env,
      ATTEST_attestAliases: JSON.stringify(['expect']),
    },
  })
}

export async function globalSetupAttest(project: TestProject) {
  // TODO: vitest level config
  // TODO: workspace support?
  if (!process.env.VITEST_ATTEST) {
    return
  }

  await setup()
  project.onTestsRerun(() => setup())
}
