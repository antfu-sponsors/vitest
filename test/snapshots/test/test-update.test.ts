import fs from 'node:fs/promises'
import { join } from 'node:path'
import { expect, test } from 'vitest'

const dir = join(import.meta.dirname, '../test-update')
const files = await fs.readdir(dir)
test.for(files)('%s', async (file) => {
  const content = await fs.readFile(join(dir, file), 'utf8')
  expect(content).toMatchSnapshot()
})
