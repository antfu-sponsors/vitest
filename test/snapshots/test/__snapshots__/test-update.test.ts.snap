// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`inline.test.js 1`] = `
"import { describe, expect, test } from 'vitest'

// when snapshots are generated Vitest reruns \`toMatchInlineSnapshot\` checks
// please, don't commit generated snapshots
describe('snapshots are generated in correct order', async () => {
  test('first snapshot', () => {
    expect({ foo: ['bar'] }).toMatchInlineSnapshot(\`
      Object {
        "foo": Array [
          "bar",
        ],
      }
    \`)
  })

  test('second snapshot', () => {
    expect({ foo: ['zed'] }).toMatchInlineSnapshot(\`
      Object {
        "foo": Array [
          "zed",
        ],
      }
    \`)
  })
})

describe('snapshots with properties', () => {
  test('without snapshot', () => {
    expect({ foo: 'bar' }).toMatchInlineSnapshot({ foo: expect.any(String) }, \`
      Object {
        "foo": Any<String>,
      }
    \`)
  })

  test('with snapshot', () => {
    expect({ first: { second: { foo: 'bar' } } }).toMatchInlineSnapshot({ first: { second: { foo: expect.any(String) } } }, \`
      Object {
        "first": Object {
          "second": Object {
            "foo": Any<String>,
          },
        },
      }
    \`)
  })

  test('mixed with and without snapshot', () => {
    expect({ first: { second: { foo: 'bar' } } }).toMatchInlineSnapshot({ first: { second: { foo: expect.any(String) } } }, \`
      Object {
        "first": Object {
          "second": Object {
            "foo": Any<String>,
          },
        },
      }
    \`)

    expect({ first: { second: { foo: 'zed' } } }).toMatchInlineSnapshot(\`
      Object {
        "first": Object {
          "second": Object {
            "foo": "zed",
          },
        },
      }
    \`)
  })
})
"
`;

exports[`inline-concurrent.test.js 1`] = `
"import { it } from 'vitest'

it.concurrent('1st', ({ expect }) => {
  expect('hi1').toMatchInlineSnapshot(\`"hi1"\`)
})

it.concurrent('2nd', ({ expect }) => {
  expect('hi2').toMatchInlineSnapshot(\`"hi2"\`)
})

it.concurrent('3rd', ({ expect }) => {
  expect('hi3').toMatchInlineSnapshot(\`"hi3"\`)
})
"
`;
